test/e2e/syncer/syncer_test.go

```diff
@@ -229,6 +247,69 @@ func TestSyncerLifecycle(t *testing.T) {
 		return true
 	}, wait.ForeverTestTimeout, time.Millisecond*100, "downstream deployment %s/%s was not synced", downstreamNamespaceName, upstreamDeployment.Name)
 
+	// Add a virtual Finalizer to the deployment and update it.
+	t.Logf("Adding a virtual finalizer to the upstream deployment %s/%s in order to simulate an external controller", upstreamNamespace.Name, upstreamDeployment.Name)
+	upstreamDeployment, err = upstreamKubeClient.AppsV1().Deployments(upstreamNamespace.Name).Get(ctx, upstreamDeployment.Name, metav1.GetOptions{})
+	require.NoError(t, err)
+
+	upstreamDeployment = upstreamDeployment.DeepCopy()
+	annotations := upstreamDeployment.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string, 0)
+	}
+	annotations["finalizers.workload.kcp.dev/"+syncerFixture.SyncerConfig.WorkloadClusterName] = "external-controller-finalizer"
+	upstreamDeployment.SetAnnotations(annotations)
+	_, err = upstreamKubeClient.AppsV1().Deployments(upstreamNamespace.Name).Update(ctx, upstreamDeployment, metav1.UpdateOptions{})
+	require.NoError(t, err)
+
+	t.Logf("Deleting upstream deployment %s/%s", upstreamNamespace.Name, upstreamDeployment.Name)
+	err = upstreamKubeClient.AppsV1().Deployments(upstreamNamespace.Name).Delete(ctx, upstreamDeployment.Name, metav1.DeleteOptions{GracePeriodSeconds: pointer.Int64(0)})
+	require.NoError(t, err)
+
+	t.Logf("Checking if upstream deployment %s/%s is getting deleted, shouldn't as there is a virtual finalizer", upstreamNamespace.Name, upstreamDeployment.Name)
```

Comment:
Isn't it the real syncer finalizer that prevents the upstream object to be deleted (and not the virtual) ?

Reply:
Without the virtual finalizer, the syncer will proceed to remove the object downstream and its finalizer. From the k8s point of view, yes it's the finalizer... but from the point of view of the user/external controller, it's the virtual finalizer.